bitbots_convenience_frames/src/convenience_frames.cpp

```diff
@@ -0,0 +1,160 @@
+#include <ros/ros.h>
+#include <sensor_msgs/Imu.h>
+#include <std_msgs/Char.h>
+#include <std_msgs/Time.h>
+#include <tf2_ros/transform_broadcaster.h>
+#include <tf2_ros/transform_listener.h>
+#include <tf2_geometry_msgs/tf2_geometry_msgs.h>
+#include <tf2/utils.h>
+
+
+class ConvenienceFramesBroadcaster
+{
+public:
+    ConvenienceFramesBroadcaster();
+private:
+    geometry_msgs::TransformStamped tf;
+    tf2_ros::Buffer tfBuffer;
+    tf2_ros::TransformListener tfListener;
+    bool is_left_support, got_support_foot;
+    void supportFootCallback(const std_msgs::Char msg);
+};
+
+ConvenienceFramesBroadcaster::ConvenienceFramesBroadcaster() : tfBuffer(ros::Duration(10.0)),  tfListener(tfBuffer)
+{
+    //setup tf listener and broadcaster as class members
+    
+    ros::NodeHandle n("~");
+    got_support_foot = false;
+    ros::Subscriber walking_support_foot_subscriber = n.subscribe("/walk_support_state", 1, &ConvenienceFramesBroadcaster::supportFootCallback, this, ros::TransportHints().tcpNoDelay());
+    ros::Subscriber dynamic_kick_support_foot_subscriber = n.subscribe("/dynamic_kick_support_state", 1, &ConvenienceFramesBroadcaster::supportFootCallback, this, ros::TransportHints().tcpNoDelay());
+
+    tf = geometry_msgs::TransformStamped();
+    
+    static tf2_ros::TransformBroadcaster br;
+    ros::Rate r(100.0);
```

Comment:
what do you mean? which rate is appropriate?

Reply:
As an example: the publishing frequency of the robot_state_publisher is 50Hz by default (we may want to even increase this since 20ms is quite a bit of time to interpolate correctly). We might want to evaluate the performance of this code at different publishing rates and see if it makes a real difference. We need to be a bit careful though since when we have more traffic on the tf topic, other code will also slow down. I think its really hard to estimate how much it slows down the code without measurements.

Reply:
the publishing rate of the robot state publisher changed